File: pulsar/java/src/main/java/com/examples/hello/pulsar/functions/FunctionFactory.java
package com.examples.hello.pulsar.functions;

import com.examples.hello.pulsar.weather.conditions.Condition;
import com.examples.hello.pulsar.services.PulsarFactory;
import io.micrometer.core.instrument.MeterRegistry;
import io.micronaut.context.annotation.Factory;
import io.micronaut.context.annotation.Value;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import lombok.SneakyThrows;
import org.apache.pulsar.client.api.Schema;
import org.apache.pulsar.functions.api.Context;

import java.util.Collections;

@Factory
public class FunctionFactory {
  @Inject
  PulsarFactory factory;

  @Value("${pulsar.function.name}")
  String name;
  @Value("${pulsar.function.input}")
  String input;
  @Value("${pulsar.function.output}")
  String output;
  @Inject
  MeterRegistry registry;

  @Singleton
  public FunctionRunnable service() {
    return FunctionRunner.<Condition, Float>builder()
      .consumer(factory.consumerOf(name, input, Condition.class))
      .function(new FilterTemperatureFunction())
      .producer(factory.producerOf(name, output, Float.class))
      .context(context())
      .build();
  }

  @SneakyThrows
  private Context context() {
    return SimpleFunctionContext.builder()
      .tenant(factory.getTenant())
      .namespace(factory.getNamespace())
      .pulsarAdmin(factory.admin())
      .inputTopics(Collections.singleton(input))
      .functionName(name)
      .registry(registry)
      .outputSchema(Schema.FLOAT)
      .build();
  }
}

File: pulsar/kotlin/src/main/kotlin/com/example/hello/pulsar/functions/FunctionFactory.kt
package com.example.hello.pulsar.functions

import com.example.hello.pulsar.services.PulsarFactory
import com.example.hello.pulsar.services.PulsarFactory.Companion.shared
import io.micronaut.context.annotation.Factory
import io.micronaut.context.annotation.Value
import jakarta.inject.Inject
import jakarta.inject.Singleton
import org.apache.pulsar.client.impl.schema.AvroSchema

@Factory
class FunctionFactory(
    @Inject private val factory: PulsarFactory,
    @Value("\${pulsar.function.name}") val functionName: String,
    @Value("\${pulsar.function.input}") val inputTopic: String,
    @Value("\${pulsar.function.output}") val outputTopic: String
) {

  @Singleton
  fun function(): FunctionRunnable {
    val func = FilterTemperatureFunction()

    val source = factory.weatherConsumer(inputTopic)
      .shared(functionName)
      .consumerName(functionName)
      .subscribe()

    val sink = factory.producer(AvroSchema.of(Float::class.java))
      .topic("${factory.tenant}/${factory.namespace}/$outputTopic")
      .also { factory.ensureTopic(outputTopic) }
      .create()

    val context = FunctionContext(
        factory.tenant, factory.namespace,
        functionName, inputTopic, outputTopic
    )
    return FunctionRunner(source, func, context, sink)
  }
}
